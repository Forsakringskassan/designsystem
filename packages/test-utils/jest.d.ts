export * from "./dist/types/jest";
