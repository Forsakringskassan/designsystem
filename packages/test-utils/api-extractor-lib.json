{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/lib.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/lib.d.ts"
    }
}
