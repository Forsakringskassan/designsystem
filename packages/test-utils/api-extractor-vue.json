{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/vue/index.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/vue.d.ts"
    }
}
