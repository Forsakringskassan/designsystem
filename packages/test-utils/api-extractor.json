{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "tsdocMetadata": {
        "tsdocMetadataFilePath": "<projectFolder>/dist/types/tsdoc-metadata.json"
    },
    "dtsRollup": {
        "enabled": false
    }
}
