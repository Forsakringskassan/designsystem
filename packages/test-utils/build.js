const fs = require("node:fs/promises");
const path = require("node:path");
const esbuild = require("esbuild");

const type = {
    cjs: "commonjs",
    esm: "module",
};

async function build() {
    for (const format of ["cjs", "esm"]) {
        const outdir = `dist/${format}`;
        const common = {
            logLevel: "info",
            sourcemap: true,
            external: ["vue", "@fkui/*"],
            bundle: true,
            platform: "node",
            format,
            target: "node20",
            outdir,
        };
        await esbuild.build({
            ...common,
            entryPoints: ["src/lib.ts", "src/jest.ts"],
        });
        await esbuild.build({
            ...common,
            entryPoints: [{ in: "src/vue/index.ts", out: "vue" }],
            define: {
                "process.env.NODE_ENV": JSON.stringify("development"),
            },
        });
        const pkg = JSON.stringify({ type: type[format] }, null, 2);
        await fs.writeFile(path.join(outdir, "package.json"), pkg, "utf-8");
    }
}

build().catch((err) => {
    console.error(err);
});
