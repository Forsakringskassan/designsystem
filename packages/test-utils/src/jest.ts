/* imported for side-effects only */
import "./matchers";

export * from "./lib";
