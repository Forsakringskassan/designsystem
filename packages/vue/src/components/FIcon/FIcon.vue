<script lang="ts">
import { defineComponent, type PropType } from "vue";

const Flip = ["horizontal", "vertical"];
const Rotate = ["90", "180", "270"];

export default defineComponent({
    name: "FIcon",
    inheritAttrs: false,
    props: {
        /**
         * Icon name.
         */
        name: {
            type: String,
            required: true,
        },
        /**
         * fk-icon library
         */
        library: {
            type: String,
            required: false,
            default: "f",
        },
        /**
         * Flip icon horizontally or vertically.
         *
         * Must be set to one of:
         *
         * - `horizontal`
         * - `vertical`
         */
        flip: {
            type: String as PropType<string>,
            default: null,
            required: false,
            validator(value: string): boolean {
                return Flip.includes(value);
            },
        },
        /**
         * Rotate icon.
         *
         * Must be set to one of:
         *
         * - `90`
         * - `180`
         * - `270`
         */
        rotate: {
            type: String as PropType<string>,
            default: null,
            required: false,
            validator(value: string): boolean {
                return Rotate.includes(value);
            },
        },
    },
    computed: {
        spriteKey(): string {
            return `${this.library}-icon-${this.name}`;
        },
        spriteId(): string {
            return `#${this.spriteKey}`;
        },
        modifiers(): string[] {
            const classes = [];
            if (this.flip) {
                classes.push(`icon--flip-${this.flip}`);
            }
            if (this.rotate) {
                classes.push(`icon--rotate-${this.rotate}`);
            }
            return classes;
        },
        ariaHidden(): "true" | undefined {
            const slotUsed = Boolean(this.$slots.default);
            const ariaLabel = this.$attrs["aria-label"] !== undefined;
            const ariaLabelledby = this.$attrs["aria-labelledby"] !== undefined;
            const ariaDescription = this.$attrs["aria-description"] !== undefined;
            const ariaDescribedby = this.$attrs["aria-describedby"] !== undefined;
            const hasText = slotUsed || ariaLabel || ariaLabelledby || ariaDescription || ariaDescribedby;
            return hasText ? undefined : "true";
        },
    },
});
</script>

<template>
    <!-- [html-validate-disable-block fkui/prefer-ficon -- this is the FIcon component]-->
    <svg v-bind="$attrs" focusable="false" class="icon" :class="[spriteKey, ...modifiers]" :aria-hidden="ariaHidden">
        <slot></slot>
        <use :href="spriteId" />
    </svg>
</template>
