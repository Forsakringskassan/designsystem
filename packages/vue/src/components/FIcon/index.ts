export { default as FIcon } from "./FIcon.vue";
