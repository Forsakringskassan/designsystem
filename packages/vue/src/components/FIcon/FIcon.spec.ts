import { VueWrapper, mount } from "@vue/test-utils";
import {
    cjsResolver,
    FileSystemConfigLoader,
    HtmlValidate,
} from "html-validate/node";
import FIcon from "./FIcon.vue";
import "html-validate/jest";

const loader = new FileSystemConfigLoader([cjsResolver()], {
    extends: [
        "html-validate:recommended",
        "html-validate-vue:recommended",
        "@fkui/vue:recommended",
    ],
    plugins: ["<rootDir>/htmlvalidate", "html-validate-vue"],
});
const htmlvalidate = new HtmlValidate(loader);

function createWrapper({
    props = {},
    slots = {},
    attrs = {},
} = {}): VueWrapper {
    return mount(FIcon, {
        attrs: { ...attrs },
        props: { name: "my-icon", ...props },
        slots: { ...slots },
    });
}

describe("snapshots", () => {
    it("should match snapshot and use prop name to set icon", () => {
        const wrapper = createWrapper();

        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot when passing content to default slot", () => {
        const wrapper = createWrapper({
            slots: {
                default: /* HTML */ ` <title>FIcon test information</title> `,
            },
        });

        expect(wrapper.element).toMatchSnapshot();
    });
});

it("should set css classes", () => {
    expect.assertions(1);
    const wrapper = createWrapper();
    const classes = Array.from(wrapper.element.classList);
    expect(classes).toEqual(["icon", "f-icon-my-icon"]);
});

describe("props", () => {
    describe("flip", () => {
        describe("should set correct css class", () => {
            it.each`
                value           | expected
                ${"horizontal"} | ${"icon--flip-horizontal"}
                ${"vertical"}   | ${"icon--flip-vertical"}
            `("$value", ({ value, expected }) => {
                const wrapper = createWrapper({
                    props: {
                        flip: value,
                    },
                });
                const classes = wrapper.element.classList;
                expect(classes).toContain(expected);
            });
        });

        describe("html-validate", () => {
            it("should allow valid values", () => {
                expect.assertions(1);
                const markup = /* HTML */ `
                    <f-icon name="my-icon" flip="horizontal"></f-icon>
                    <f-icon name="my-icon" flip="vertical"></f-icon>
                `;
                const report = htmlvalidate.validateString(markup, "foo.html");
                expect(report).toBeValid();
            });

            it("should not allow invalid values", () => {
                expect.assertions(2);
                const markup = /* HTML */ `
                    <f-icon name="my-icon" flip="foo"></f-icon>
                `;
                const report = htmlvalidate.validateString(markup, "file.html");
                expect(report).toBeInvalid();
                expect(report).toHaveError(
                    "attribute-allowed-values",
                    'Attribute "flip" has invalid value "foo"',
                );
            });
        });
    });

    describe("rotate", () => {
        describe("should set correct css class", () => {
            it.each`
                value    | expected
                ${"90"}  | ${"icon--rotate-90"}
                ${"180"} | ${"icon--rotate-180"}
                ${"270"} | ${"icon--rotate-270"}
            `("$value", ({ value, expected }) => {
                const wrapper = createWrapper({
                    props: {
                        rotate: value,
                    },
                });
                const classes = wrapper.element.classList;
                expect(classes).toContain(expected);
            });
        });

        describe("html-validate", () => {
            it("should allow valid values", async () => {
                expect.assertions(1);
                const markup = /* HTML */ `
                    <f-icon name="my-icon" rotate="90"></f-icon>
                    <f-icon name="my-icon" rotate="180"></f-icon>
                    <f-icon name="my-icon" rotate="270"></f-icon>
                `;
                const report = await htmlvalidate.validateString(markup);
                expect(report).toBeValid();
            });

            it("should not allow arbitrary degrees", async () => {
                expect.assertions(2);
                const markup = /* HTML */ `
                    <f-icon name="my-icon" rotate="42"></f-icon>
                `;
                const report = await htmlvalidate.validateString(markup);
                expect(report).toBeInvalid();
                expect(report).toMatchInlineCodeframe(`
                    "error: Attribute "rotate" has invalid value "42" (attribute-allowed-values) at inline:2:52:
                      1 |
                    > 2 |                     <f-icon name="my-icon" rotate="42"></f-icon>
                        |                                                    ^^
                      3 |
                    Selector: f-icon"
                `);
            });

            it("should not allow invalid values", async () => {
                expect.assertions(2);
                const markup = /* HTML */ `
                    <f-icon name="my-icon" rotate="foo"></f-icon>
                `;
                const report = await htmlvalidate.validateString(markup);
                expect(report).toBeInvalid();
                expect(report).toMatchInlineCodeframe(`
                    "error: Attribute "rotate" has invalid value "foo" (attribute-allowed-values) at inline:2:52:
                      1 |
                    > 2 |                     <f-icon name="my-icon" rotate="foo"></f-icon>
                        |                                                    ^^^
                      3 |
                    Selector: f-icon"
                `);
            });
        });
    });
});

describe("aria-hidden", () => {
    it("should be true by default", () => {
        expect.assertions(1);
        const wrapper = createWrapper();
        const icon = wrapper.get(".icon");
        expect(icon.attributes("aria-hidden")).toBe("true");
    });

    it("should be undefined if slot is used", () => {
        expect.assertions(1);
        const wrapper = createWrapper({
            slots: {
                default: /* HTML */ ` <title>FIcon description</title> `,
            },
        });
        const icon = wrapper.get(".icon");
        expect(icon.attributes("aria-hidden")).toBeUndefined();
    });

    it("should be undefined if aria-label is set", () => {
        expect.assertions(1);
        const TestComponent = {
            components: { FIcon },
            template: /* HTML */ `
                <f-icon name="my-icon" aria-label="My awesome label"></f-icon>
            `,
        };
        const wrapper = mount(TestComponent);
        const icon = wrapper.get(".icon");
        expect(icon.attributes("aria-hidden")).toBeUndefined();
    });

    it("should be undefined if aria-labelledby is set", () => {
        expect.assertions(1);
        const TestComponent = {
            components: { FIcon },
            template: /* HTML */ `
                <div>
                    <f-icon name="my-icon" aria-labelledby="my-label"></f-icon>
                    <span id="my-label"> My awesome label </span>
                </div>
            `,
        };
        const wrapper = mount(TestComponent);
        const icon = wrapper.get(".icon");
        expect(icon.attributes("aria-hidden")).toBeUndefined();
    });

    it("should be undefined if aria-description is set", () => {
        expect.assertions(1);
        const TestComponent = {
            components: { FIcon },
            template: /* HTML */ `
                <div>
                    <f-icon
                        name="my-icon"
                        aria-description="My awesome description"
                    ></f-icon>
                </div>
            `,
        };
        const wrapper = mount(TestComponent);
        const icon = wrapper.get(".icon");
        expect(icon.attributes("aria-hidden")).toBeUndefined();
    });

    it("should be undefined if aria-describedby is set", () => {
        expect.assertions(1);
        const TestComponent = {
            components: { FIcon },
            template: /* HTML */ `
                <div>
                    <f-icon
                        name="my-icon"
                        aria-describedby="my-description"
                    ></f-icon>
                    <span id="my-description"> My awesome description </span>
                </div>
            `,
        };
        const wrapper = mount(TestComponent);
        const icon = wrapper.get(".icon");
        expect(icon.attributes("aria-hidden")).toBeUndefined();
    });
});
