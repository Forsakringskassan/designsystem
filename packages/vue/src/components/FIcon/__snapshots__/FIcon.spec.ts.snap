// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match snapshot and use prop name to set icon 1`] = `
<svg
  aria-hidden="true"
  class="icon f-icon-my-icon"
  focusable="false"
>
  
  
  <use
    href="#f-icon-my-icon"
  />
</svg>
`;

exports[`snapshots should match snapshot when passing content to default slot 1`] = `
<svg
  class="icon f-icon-my-icon"
  focusable="false"
>
  
  <title>
    FIcon test information
  </title>
  
  <use
    href="#f-icon-my-icon"
  />
</svg>
`;
